### PR TITLE
Add SessionDisconnectEvent

### PR DESCRIPTION
**Pull request by onebeastchris**
This PR adds a SessionDisconnectEvent. While also present in all server api's (paper, velocity, fabric etc), this one can be used by Extensions specifically to e.g. log which disconnect reasons occur, or to change disconnect reasons. 

The way how disconnects are handled currently are quite... complicated, since there are 4 possible ways where session.disconnect is called normally:
- when the java server disconnects the session in either [JavaLoginDisconnectTranslator](https://github.com/GeyserMC/Geyser/blob/master/core/src/main/java/org/geysermc/geyser/translator/protocol/java/JavaLoginDisconnectTranslator.java), or the [JavaDisconnectTranslator](https://github.com/GeyserMC/Geyser/blob/master/core/src/main/java/org/geysermc/geyser/translator/protocol/java/JavaDisconnectTranslator.java).
- when the bedrock connection is closed, GeyserSession.disconnect() is called to close the java server connection: [here](https://github.com/GeyserMC/Geyser/blob/master/core/src/main/java/org/geysermc/geyser/network/UpstreamPacketHandler.java#L121-L128)
- LoginEncryptionUtils/UpstreamPacketHandler for stuff like invalid chain/unsupported version/etc
- Finally, by the [SessionManager](https://github.com/GeyserMC/Geyser/blob/master/core/src/main/java/org/geysermc/geyser/session/SessionManager.java#L92-L97). This however is broken; even though upstream.disconnect is called the client never gets the disconnect message and times out.

This is all fine and well, but leaves the question of where to put this new SessionDisconnectEvent to avoid it firing twice.
My initial thought would be to add it to [UpstreamSession](https://github.com/onebeastchris/Geyser/blob/leave-event/core/src/main/java/org/geysermc/geyser/session/UpstreamSession.java#L63-L71)'s disconnect method; but that has a few problems: 
- It isnt called when the Bedrock client disconnects (while GeyserSession.disconnect() is called; upstream.isClosed prohibits calling upstream.disconnect()
- Additionally to that, when the emulated Java client disconnects, it also calls upstream.disconnect() [here](https://github.com/GeyserMC/Geyser/blob/master/core/src/main/java/org/geysermc/geyser/session/GeyserSession.java#L1031). 

My proposed fix that I've implemented in https://github.com/GeyserMC/Geyser/pull/4052/commits/2abf1af26b2acbab2f6e26427c4a0eb92b49ca5f :
- fire the SessionDisconnectEvent in GeyserSession.disconnect(), since that is already "gated" to not fire twice (@Camotoy , you were right; but atleast i can fix some other minor issues while at it :p)

Other minor tweaks:
- Stop the DisconnectedEvent from calling upstream.disconnect() directly and let it call GeyserSession.disconnect() instead (Since otherwise; the SessionDisconnectEvent fired in GeyserSession.disconnect() can not change the disconnect message). 
- There is no need to call downstream.disconnect() if the downstream session is already closed in GeyserSession.disconnect()
- add a few comments
- Imo there is no need to print a full stacktrace when a connection is impossible due to e.g. the server being offline; gated that behind debug mode
- /geyser reload now properly kicks players (added a 10ms wait before executing reload.. while hacky, i dont see any other solution here). /geyser stop still leaves players hanging to time out; not sure how to do that 
-> From testing; it looks like upstream.disconnect() is called for each player; but is then somehow interrupted? Not sure there, tbh.

### CODE REVIEW

File: api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionLeaveEvent.java
Comment (Konicai) [inline api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionLeaveEvent.java:60]: Add newline to EOF

**Review by Konicai — APPROVED** [review]
Looks good to me. Lets get a project lead review

File: api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionLeaveEvent.java
Comment (Tim203) [inline api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionLeaveEvent.java:35]: I'd call it `SessionDisconnectEvent`. Leaving sounds intentional and initiated from the client, but it's called for all scenarios.

File: core/src/main/java/org/geysermc/geyser/session/GeyserSession.java
Comment (Tim203) [inline core/src/main/java/org/geysermc/geyser/session/GeyserSession.java:1099]:
```suggestion
        geyser.eventBus().fire(leaveEvent);
```

**Review by Tim203 — CHANGES_REQUESTED** [review]

**Comment by onebeastchris** [comment]
converting to draft for the time being, the placement of the event firing wont work like that (it can be fired twice). On hold until i find a better spot to put it

**Comment by Konicai** [comment]
Yeah, after doing debug logging for latency stuff, I realized that the code path for a disconnect can be wildly different depending on what initiated it

**Comment by Camotoy** [comment]
Just move the event calling to after `if (!closed) {` ?

File: core/src/main/java/org/geysermc/geyser/command/defaults/ReloadCommand.java
Comment (Camotoy) [inline core/src/main/java/org/geysermc/geyser/command/defaults/ReloadCommand.java:57]: Can you elaborate (in the code) on why this behavior changes things?

File: core/src/main/java/org/geysermc/geyser/session/GeyserSession.java
Comment (Camotoy) [inline core/src/main/java/org/geysermc/geyser/session/GeyserSession.java:1078]: Are we intentionally not printing out the custom disconnect reason here?

Comment (onebeastchris) [reply core/src/main/java/org/geysermc/geyser/session/GeyserSession.java:1078]: Yeah, since that could otherwise suppress errors - we dont want that xd
Not sure whether we should send the custom disconnect reason to the java server though?

File: core/src/main/java/org/geysermc/geyser/command/defaults/ReloadCommand.java
Comment (onebeastchris) [reply core/src/main/java/org/geysermc/geyser/command/defaults/ReloadCommand.java]: Added a fixme; as for why, i dont really know - in my testing, session.disconnect was called for all sessions through disconnectAll(); but that never really disconnected players.

**Review by Camotoy — APPROVED** [review]

File: api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionDisconnectEvent.java
Comment (Tim203) [inline api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionDisconnectEvent.java:35]: I think that `ConnectionDisconnectEvent` makes more sense, because a `Connection` (API term) is disconnected. Session is the internal name.
I don't think that we should add `Connection` variants and deprecate the `Session` variants of all the other events in this PR, but if that's the route we choose then this event should be `ConnectionDisconnectEvent`.
We'll have to discuss this internally.

**Review by Tim203 — CHANGES_REQUESTED** [review]
The actual code is fine, just one thing we have to discuss before this can be merged.

File: api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionDisconnectEvent.java
Comment (onebeastchris) [reply api/src/main/java/org/geysermc/geyser/api/event/bedrock/SessionDisconnectEvent.java:35]: Resolved internally - all session prefixed events indicate that they're Geyser-related, either called by Geyser for a session or requiring Geyser processing

**Review by Tim203 — APPROVED** [review]

**Pull request merged**
